Implement axis parameter on cupy.unique

andfoy opened the pull request:
Fixes #6691 

This is different from the numpy implementation, since they use structured arrays to keep track of the index changes. This implementation sorts the axis elements manually (since there's no a clear way to do it more "generic" via `argsort`)

andfoy commented on cupy/_manipulation/add_remove.py:
```diff
@@ -169,15 +174,114 @@ def unique(ar, return_index=False, return_inverse=False,
+            left = ar[idx1, i]
+            right = ar[idx2, i]
+
+            if cupy.isnan(left) and cupy.isnan(right):
```
I'm now sure which should be the behaviour regarding NaNs when `unique` is called with a given axis

leofang replied:
Maybe it's worth checking the NumPy behavior? Note that NumPy has a new kw-only arg `equal_nan`.

andfoy replied:
The NumPy behaviour is not consistent across datatypes, sometimes `NaN`s get pushed to the beginning on integral types and sometimes to the end on floating types. I'm not sure how to reconcile this two types of behaviours.

andfoy replied:
This is the output of the comparison tests when `equal_nan` is passed:

![imagen](https://user-images.githubusercontent.com/1878982/187954433-e4e73d64-7aa4-4c85-a2e3-7a09cdc4cb08.png)

For instance, `float32` types have a consistent return when they are compared against NumPy, however, other types such as `float16` and `complex` types behave differently:

## float16 - equal_nan=True
```
cupy_result = array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]], dtype=float16)

numpy_result = array([[nan,  1., nan],
       [nan,  1., nan],
       [ 2., nan,  2.]], dtype=float16)
```

As you can see, when `equal_nan=True`, then `NaNs` get pushed to the front of the result, whereas when `equal_nan=False`, they get pushed to the end. 

Given that this PR makes `NaNs` greater when `equal_nan=True`, that means that they will get pushed to the end rather to the beginning. However, as it is seen on the examples, NumPy sometimes pushes them to the front and others to the end, depending on the datatype

rgommers replied:
@andfoy some testing shows that `numpy` has a bug for `float16`, and all other dtypes seem fine:
```
>>> xp = np
>>> a = xp.array([[2, xp.nan, 2],
...               [xp.nan, 1, xp.nan],
...               [xp.nan, 1, xp.nan]])
>>> np.unique(a.astype(np.float64), axis=0)
array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]])
>>> np.unique(a.astype(np.float32), axis=0)
array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]], dtype=float32)
>>> np.unique(a.astype(np.float128), axis=0)
array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]], dtype=float128)
>>> np.unique(a.astype(np.float16), axis=0)
array([[nan,  1., nan],
       [nan,  1., nan],
       [ 2., nan,  2.]], dtype=float16)
>>> np.unique(a.astype(np.complex256), axis=0)
array([[ 2.+0.j, nan+0.j,  2.+0.j],
       [nan+0.j,  1.+0.j, nan+0.j],
       [nan+0.j,  1.+0.j, nan+0.j]], dtype=complex256)
>>> np.unique(a.astype(np.complex128), axis=0)
array([[ 2.+0.j, nan+0.j,  2.+0.j],
       [nan+0.j,  1.+0.j, nan+0.j],
       [nan+0.j,  1.+0.j, nan+0.j]])
>>> np.unique(a.astype(np.complex64), axis=0)
array([[ 2.+0.j, nan+0.j,  2.+0.j],
       [nan+0.j,  1.+0.j, nan+0.j],
       [nan+0.j,  1.+0.j, nan+0.j]], dtype=complex64)
```

CuPy does the right thing here, and NumPy has a bug specific to `float16`. That is true independent of the value of `equal_nan`. So I suggest to open a new issue on the NumPy issue tracker, and skip the `float16` tests in this PR.

The second issue is the failures for complex dtypes with `equal_nan=True`. I don't have a `cupy` install at hand, can you please post a simple reproducer like mine above?

andfoy commented:
I've just extended the `NaN` comparison (lexicographical sorting element-wise on pairs of rows of the axis, looking if `left_row < right_row`) to consider the following cases:

### `equal_nan=False` and non-complex dtype
1. If both elements are `NaN`, then the left row is marked as "lesser" and moved to the left of the result.
2. If `left_elem = NaN` and `right_elem != NaN`, then the left row is marked as "lesser" and moved to the left of the result.
3. If  `left_elem != NaN` and `right_elem = NaN`, then the left row is marked as "greater" and moved to the right of the result.

### `equal_nan=True` and non-complex dtype
1. If both elements are `NaN`, then the comparison is skipped to the next element.
2. If `left_elem = NaN` and `right_elem != NaN`: same as `equal_nan=False`.
3. If  `left_elem != NaN` and `right_elem = NaN`: same as `equal_nan=False`.

### `equal_nan=False` and complex dtype
1. If `left_elem = (NaN, NaNj)` and `right_elem = (NaN, NaNj)`, then the left row is marked as "lesser" and moved to the left of the result.
2. If `left_elem = (NaN, NaNj)` and `right_elem = (Y, NaNj)`, then the left row is marked as "lesser" and moved to the left of the result.
3. If `left_elem = (NaN, NaNj)` and `right_elem = (NaN, Wj)`, then the left row is marked as "lesser" and moved to the left of the result.
4. If `left_elem = (NaN, NaNj)` and `right_elem = (Y, Wj)`, then the left row is marked as "lesser" and moved to the left of the result.
5. If `left_elem = (X, NaNj)` and `right_elem = (NaN, NaNj)`, then the left row is marked as "greater" and moved to the right of the result.
6. If `left_elem = (X, NaNj)` and `right_elem = (Y, NaNj)`, then the comparison is made between the real parts.
7. If `left_elem = (X, NaNj)` and `right_elem = (NaN, Wj)`, then the left row is marked as "greater" and moved to the right of the result.
8. If `left_elem = (X, NaNj)` and `right_elem = (Y, Wj)`, then the comparison is made between the real parts.
9.  If `left_elem = (NaN, Kj)` and `right_elem = (NaN, NaNj)`, then the left row is marked as "greater" and moved to the right of the result.
10. If `left_elem = (NaN, Kj)` and `right_elem = (Y, NaNj)`, then the left row is marked as "lesser" and moved to the left of the result.
11. If `left_elem = (NaN, Kj)` and `right_elem = (NaN, Wj)`, then the comparison is made between the imaginary parts.
12. If `left_elem = (NaN, Kj)` and `right_elem = (Y, Wj)`,  then the left row is marked as "lesser" and moved to the left of the result.
13. If `left_elem = (X, Kj)` and `right_elem = (NaN, NaNj)`, then the left row is marked as "greater" and moved to the right of the result.
14. If `left_elem = (X, Kj)` and `right_elem = (Y, NaNj)`, then the comparison is made between the real parts.
15. If `left_elem = (X, Kj)` and `right_elem = (NaN, Wj)`, then the left row is marked as "greater" and moved to the right of the result.
16. If `left_elem = (X, Kj)` and `right_elem = (Y, Wj)`, then the comparison is performed normally.

### `equal_nan=True` and complex dtype
1. If `left_elem = (NaN, NaNj)` and `right_elem = (NaN, NaNj)`, then the comparison is skipped.
2. The other cases remain the same as `equal_nan=False`

asi1024 commented:
@toslunar Could you start a review?

toslunar reviewed (changes requested):
`equal_nan` option seems to have no effect with `axis` option (or with structured arrays) in NumPy.  It may be just a specification because I didn't find NumPy issue about it.

I expect CuPy implementation is faster than `cupy.array(numpy.unique(cupy.asnumpy(ar), axis))`.  The current implementation uses Python-level, which may be slower than NumPy's implementation.  (For some other functions in CuPy, their moderate performance is OK just because they avoid CPU-GPU sync.  This isn't the case with `unique`).

Therefore I suggest using comparison on GPU in `compare_axis_elems(idx1, idx2)` by something like `cupy.trim_zeros(ar[idx1] - ar[idx2], 'f')` for better performance on small number of large vectors, where comparison operations are relatively heavy.

A harder case is when `ar` contains large number of small vectors. I'd like to use bucket sort.  This can be out of the scope of the PR.

It's also possible that doing only pre/post-processing on GPU is justified in the sense of performance.  For example,
```python
ar_cpu = cupy.asnumpy(cupy.ascontiguousarray(cupy.moveaxis(ar, ...).reshape(...)))
_, sorted_indices = numpy.unique(ar_cpu, return_index=True, axis=0)
return cupy.take(ar, sorted_indices, axis)
```

andfoy commented:
@toslunar, thanks for your review! I've just updated the implementation to use `trim_zeros` instead of elementwise scan, I'm just missing a performance comparison between calling `unique` on CPU vs the proposed implementation

andfoy commented:
Here are the profiling results where the current implementation (using `trim_zeros` + subtraction)  is compared against pre/post-processing on GPU and then calling NumPy's `unique` function as proposed by @toslunar on https://github.com/cupy/cupy/pull/6886#pullrequestreview-1140313305. 

The comparison script corresponds to this one: https://gist.github.com/andfoy/7244295739c7636ed0e16ac87ca9d458. All the times reported consist of the maximum between the GPU and the CPU elapsed times.

As it is possible to observe, the current implementation based on a lexicographical sort of the rows of the dimension being sorted is much faster than a potential implementation based on calling `unique` on CPU and then indexing

| Size | `dtype` | GPU-only (ms) | GPU for pre/post-processing + CPU (ms) |
|:----:|:-------:|:-------------:|:-------------:|
| 3x3 | `uint8` | 0.314971 | 0.236437
| 5x5 | `uint8` | 0.288600 | 0.240530
| 10x10 | `uint8` | 0.291211 | 0.250625
| 100x100 | `uint8` | 0.369488 | 0.664650
| 500x500 | `uint8` | 0.406018 | 5.873621
| 1000x1000 | `uint8` | 0.800287 | 20.733934
| 5000x5000 | `uint8` | 9.005074 | 460.965287
| 10000x10000 | `uint8` | 34.741940 | 1857.037542
| 3x3 | `int32` | 0.336972 | 0.209263
| 5x5 | `int32` | 0.304245 | 0.234205
| 10x10 | `int32` | 0.304105 | 0.263415
| 100x100 | `int32` | 0.419373 | 0.677363
| 500x500 | `int32` | 0.572477 | 6.331609
| 1000x1000 | `int32` | 1.202842 | 22.220180
| 5000x5000 | `int32` | 16.069151 | 554.746273
| 10000x10000 | `int32` | 63.799783 | 2073.969380
| 3x3 | `int64` | 0.417643 | 0.267951
| 5x5 | `int64` | 0.401194 | 0.231783
| 10x10 | `int64` | 0.294274 | 0.253717
| 100x100 | `int64` | 0.414672 | 0.696820
| 500x500 | `int64` | 0.854722 | 6.263096
| 1000x1000 | `int64` | 2.015497 | 23.052823
| 5000x5000 | `int64` | 39.345610 | 585.807965
| 10000x10000 | `int64` | 156.096988 | 2264.345225
| 3x3 | `float32` | 0.309342 | 0.217062
| 5x5 | `float32` | 0.321131 | 0.243317
| 10x10 | `float32` | 0.313809 | 0.269831
| 100x100 | `float32` | 0.415692 | 0.674457
| 500x500 | `float32` | 0.643197 | 6.248911
| 1000x1000 | `float32` | 1.586711 | 22.506326
| 5000x5000 | `float32` | 21.515723 | 556.062853
| 10000x10000 | `float32` | 85.151700 | 2076.375974
| 3x3 | `float64` | 0.374232 | 0.216780
| 5x5 | `float64` | 0.329921 | 0.250573
| 10x10 | `float64` | 0.326812 | 0.266707
| 100x100 | `float64` | 0.465202 | 0.678642
| 500x500 | `float64` | 1.021641 | 6.314075
| 1000x1000 | `float64` | 2.320133 | 23.217900
| 5000x5000 | `float64` | 48.405550 | 588.974063
| 3x3 | `complex64` | 0.344953 | 0.221207
| 5x5 | `complex64` | 0.307549 | 0.240872
| 10x10 | `complex64` | 0.311879 | 0.257227
| 100x100 | `complex64` | 0.411276 | 0.685826
| 500x500 | `complex64` | 0.970940 | 6.654078
| 1000x1000 | `complex64` | 2.772588 | 23.736537
| 5000x5000 | `complex64` | 78.739919 | 619.654977
| 3x3 | `complex128` | 0.368012 | 0.211796
| 5x5 | `complex128` | 0.331925 | 0.245640
| 10x10 | `complex128` | 0.339596 | 0.266747
| 100x100 | `complex128` | 0.616204 | 0.713759
| 500x500 | `complex128` | 3.348005 | 6.816852
| 1000x1000 | `complex128` | 12.008427 | 34.436096

asi1024 commented:
/test mini

asi1024 commented:
/test mini

asi1024 commented:
CI failure seems unrelated to this PR.

asi1024 commented:
@andfoy Sorry for our late review. LGTM!
